bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java

```diff
@@ -0,0 +1,285 @@
+  private final ReadSampler readSampler;
+  private final Logger failedWritesLogger;
+  private final MirroringConfiguration configuration;
+  private final Connection primaryConnection;
+  private final Connection secondaryConnection;
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+  private final AtomicBoolean aborted = new AtomicBoolean(false);
+  private final boolean performWritesConcurrently;
+
+  /**
+   * The constructor called from {@link
+   * org.apache.hadoop.hbase.client.ConnectionFactory#createConnection(Configuration)} and in its
+   * many forms via reflection with this specific signature.
+   *
+   * <p>Parameters are passed down to ConnectionFactory#createConnection method, connection errors
+   * are passed back to the user.
+   */
+  public MirroringConnection(Configuration conf, boolean managed, ExecutorService pool, User user)
+      throws IOException {
+    assert !managed; // This is always-false legacy hbase parameter.
+    this.configuration = new MirroringConfiguration(conf);
+    this.mirroringTracer = new MirroringTracer();
+
+    this.primaryConnection =
+        ConnectionFactory.createConnection(this.configuration.primaryConfiguration, pool, user);
+    this.secondaryConnection =
+        ConnectionFactory.createConnection(this.configuration.secondaryConfiguration, pool, user);
+
+    if (pool == null) {
+      this.executorService = Executors.newCachedThreadPool();
+    } else {
+      this.executorService = pool;
+    }
+
+    referenceCounter = new ListenableReferenceCounter();
+    this.flowController =
+        new FlowController(
+            ReflectionConstructor.<FlowControlStrategy>construct(
+                this.configuration.mirroringOptions.flowControllerStrategyClass,
+                this.configuration.mirroringOptions));
+    this.mismatchDetector =
+        ReflectionConstructor.construct(
+            this.configuration.mirroringOptions.mismatchDetectorClass, this.mirroringTracer);
+
+    this.failedWritesLogger =
+        new Logger(
+            ReflectionConstructor.<Appender>construct(
+                this.configuration.mirroringOptions.writeErrorLogAppenderClass,
+                Configuration.class,
+                this.configuration),
+            ReflectionConstructor.<Serializer>construct(
+                this.configuration.mirroringOptions.writeErrorLogSerializerClass));
+
+    final SecondaryWriteErrorConsumer secondaryWriteErrorConsumer =
+        ReflectionConstructor.construct(
+            this.configuration.mirroringOptions.writeErrorConsumerClass, this.failedWritesLogger);
+
+    this.secondaryWriteErrorConsumer =
+        new SecondaryWriteErrorConsumerWithMetrics(
+            this.mirroringTracer, secondaryWriteErrorConsumer);
+    this.readSampler = new ReadSampler(this.configuration.mirroringOptions.readSamplingRate);
+    this.performWritesConcurrently = this.configuration.mirroringOptions.performWritesConcurrently;
```

Comment:
ping

Reply:
I have thought that description of config parameter would be enough. I'll add the following comment to that field:
```
  /**
   * Enables concurrent writes mode. Should always be enabled together with {@link
   * #waitForSecondaryWrites}.
   *
   * <p>In this mode some of the writes ({@link Put}s, {@link
   * org.apache.hadoop.hbase.client.Delete}s and {@link
   * org.apache.hadoop.hbase.client.RowMutations}) performed using {@link
   * org.apache.hadoop.hbase.client.Table} API will be performed concurrently and {@link
   * com.google.cloud.bigtable.mirroring.core.bufferedmutator.ConcurrentMirroringBufferedMutator}
   * instances will be returned by {@link
   * org.apache.hadoop.hbase.client.Connection#getBufferedMutator}. Moreover, all operations in this
   * mode wait for secondary database operation to finish before returning to the user (because
   * {@link #waitForSecondaryWrites} should be set).
   */
```